src/NMonad/Main.hs

```diff
@@ -15,5 +27,5 @@ nmonad cfg = do
 
 mainLoop :: N ()
 mainLoop = forever $ do
-  (dbusNotification, responseVar) <- asks globalMailbox >>= liftIO . takeMVar
+  (dbusNotification, responseVar) <- asks (view globalMailbox) >>= liftIO . takeMVar
```

Comment:
What is the difference between using and not using `view` here?

Reply:
The changes in this PR make it so Template Haskell is used to generate lenses and disable field selectors in `NMonad.Core`. I'll try to summarize what this means. Take for instance the following product type defined using record syntax:

```haskell
data Person = Person
  { name :: String
  , age :: Int
  }
```

Default behavior with GHC is to automatically create top level symbols `name :: Person -> String` and `age :: Person -> Int` used to select corresponding fields from a given `Person`. **These are called field selectors.**

The first problem this behavior introduces is you are no longer able to create types with records of the same name in a single module. This is because this goes beyond name shadowing — these are implicit top level declarations that result in a compilation error. Take the following example:

```haskell
data Person = Person
  { name :: String
  , age :: Int
  } deriving (Show)

data Animal = Animal
  { name :: String
  , age :: Int
  } deriving (Show)

main :: IO ()
main = print $ Person "John" 30
```

Trying to compile the module above will result in the following message:

```
[1 of 1] Compiling Main             ( records.hs, records.o )

records.hs:7:5: error:
    Multiple declarations of ‘name’
    Declared at: records.hs:2:5
                 records.hs:7:5
  |
7 |   { name :: String
  |     ^^^^

records.hs:8:5: error:
    Multiple declarations of ‘age’
    Declared at: records.hs:3:5
                 records.hs:8:5
  |
8 |   , age :: Int
  |     ^^^
```

It is possible to curb this limitation with [the `DisambiguateRecordFields` pragma](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/disambiguate_record_fields.html#extension-DisambiguateRecordFields), but we're trying to define lenses here, not field selectors.

Lenses are also known as Semantic Editor Combinators, vaguely defined as `type SEC s t a b = (a -> b) -> s -> t`. All that Template Haskell does is act as a Haskell preprocessor, simple metaprogramming the likes of which we have all see in C and C++. It cannot integrate lenses with GHC syntax-like functionality (like field selectors that can be disambiguated), so there is no alternative to `DisambiguateRecordFields` for lenses.

[This presentation by Edward Kmett,](https://www.youtube.com/watch?v=cefnmjtAolY) the author of the `lenses` module, does a much better job than I ever could at explaining semantic editor combinators and lenses. He does an incredible job at constructing the concept of optics and explaining the benefits of lenses over field selectors.

So let me address your question.

`globalMailbox`, after these changes, is no longer a field selector, but rather a lens (SEC), which we're using here as a getter. So we _have_ to change how we use it, and the way to do so is [`Control.Lens.view`.](https://hackage.haskell.org/package/lens-5.1.1/docs/Control-Lens-Getter.html#g:3)

Hope that helps!

Reply:
Helped for sure. 100% percent clear! thank you, sir!